File: packages/quran-engine-js/src/isnad.js
// @ts-check
/**
 * The chains of transmission (isnād) of the Ten Readings.
 *
 * From the Prophet ﷺ down through the Companions who learned from him, the Successors who
 * taught each imam, the imam himself, the links between him and each of his two narrators,
 * and the students who carried each narration on.
 *
 * The links are the standard ones of the classical record: Ibn al-Jazarī's *al-Nashr* and
 * *Ghāyat al-Nihāyah*, al-Dānī's *al-Taysīr*, and the ṭuruq of al-Shāṭibiyyah and al-Durrah.
 * Where a narrator's students are not listed with confidence, the layer is simply absent.
 *
 * A chain is returned as LAYERS, one per generation, top (the Prophet ﷺ) to bottom, which is
 * how a consumer draws it: a row per layer, connected downward.
 *
 * Riwayah keys are the canonical tags this engine uses elsewhere ("Warsh an Nafi"), and imam
 * keys are the imam's short name ("Nafi"). `chain()` takes either and works out which it has.
 *
 * @typedef {Object} IsnadNode
 * @property {string} name
 * @property {string} arabic
 * @property {string} detail  the death year, e.g. "d. 117 AH"
 * @property {"prophet"|"companion"|"successor"|"imam"|"link"|"narrator"|"student"} role
 *
 * @typedef {Object} IsnadLayer
 * @property {string} title  e.g. "THE COMPANIONS"
 * @property {IsnadNode[]} nodes
 *
 * @typedef {Object} ImamChain
 * @property {IsnadNode[]} teachers
 * @property {IsnadNode[]} companions
 *
 * @typedef {Object} NarratorChain
 * @property {string} imam          the imam this narration comes from
 * @property {IsnadNode[]} links     empty where the narrator read on the imam himself
 * @property {IsnadNode[]} students
 */

export class Isnad {
  /**
   * @param {Object} [data] parsed data/isnad.json
   * @param {IsnadNode} [data.prophet]
   * @param {IsnadNode[]} [data.companions]
   * @param {Record<string, ImamChain>} [data.imams]
   * @param {Record<string, NarratorChain>} [data.narrators]
   */
  constructor(data = {}) {
    this._prophet = data.prophet ?? null;
    this._companions = data.companions ?? [];
    this._imams = data.imams ?? {};
    this._narrators = data.narrators ?? {};
  }

  get isLoaded() {
    return Object.keys(this._imams).length > 0;
  }

  /** The Prophet ﷺ, the head of every chain. @returns {IsnadNode|null} */
  prophet() {
    return this._prophet;
  }

  /** The thirteen Companions the readings are transmitted from. @returns {IsnadNode[]} */
  companions() {
    return this._companions;
  }

  /** The ten imams' keys, e.g. "Nafi". @returns {string[]} */
  imamKeys() {
    return Object.keys(this._imams);
  }

  /** The twenty riwayah tags, e.g. "Warsh an Nafi". @returns {string[]} */
  narratorKeys() {
    return Object.keys(this._narrators);
  }

  /** @param {string} imam @returns {ImamChain|null} */
  imam(imam) {
    return this._imams[imam] ?? null;
  }

  /** @param {string} riwayah @returns {NarratorChain|null} */
  narrator(riwayah) {
    return this._narrators[riwayah] ?? null;
  }

  /**
   * Whether a narrator read on his imam himself, with nobody between them.
   * @param {string} riwayah @returns {boolean}
   */
  readsDirectly(riwayah) {
    const chain = this._narrators[riwayah];
    return !!chain && chain.links.length === 0;
  }

  /**
   * The imam a riwayah comes from. Read from the corpus, NOT parsed off the tag: four tags name
   * the imam in the Arabic genitive ("ad-Duri an Abi Amr") while his key is the nominative ("Abu
   * Amr"), so splitting on " an " would resolve those four to nothing.
   * @param {string} riwayah @returns {string|null}
   */
  imamOf(riwayah) {
    const imam = this._narrators[riwayah]?.imam;
    return imam && this._imams[imam] ? imam : null;
  }

  /**
   * The layers above any imam: the Prophet ﷺ, the Companions his teachers read on, and those
   * teachers. Shared by both `chain` forms, since every chain runs through them.
   * @param {string} imam @returns {IsnadLayer[]}
   */
  topLayers(imam) {
    const chain = this._imams[imam];
    if (!chain) return [];
    /** @type {IsnadLayer[]} */
    const layers = [];
    if (this._prophet) layers.push({ title: "THE PROPHET", nodes: [this._prophet] });
    if (chain.companions.length) layers.push({ title: "THE COMPANIONS", nodes: chain.companions });
    if (chain.teachers.length) {
      layers.push({ title: chain.teachers.length === 1 ? "HIS TEACHER" : "HIS TEACHERS", nodes: chain.teachers });
    }
    return layers;
  }

  /**
   * A whole chain as layers. Give it a riwayah tag ("Warsh an Nafi") for one narration's chain,
   * or an imam key ("Nafi") for the reading's, which ends at his narrators.
   * @param {string} key @returns {IsnadLayer[]}
   */
  chain(key) {
    const k = String(key ?? "").trim();
    if (this._imams[k]) return this._chainOfImam(k);
    if (this._narrators[k]) return this._chainOfNarrator(k);
    return [];
  }

  /** @param {string} imam @returns {IsnadLayer[]} */
  _chainOfImam(imam) {
    const layers = this.topLayers(imam);
    if (!layers.length) return [];
    layers.push({ title: "THE IMAM", nodes: [{ name: imam, arabic: "", detail: "", role: "imam" }] });
    const narrators = this.narratorKeys()
      .filter((tag) => this.imamOf(tag) === imam)
      .map((tag) => ({ name: tag.slice(0, tag.indexOf(" an ")), arabic: "", detail: "", role: /** @type {const} */ ("narrator") }));
    if (narrators.length) layers.push({ title: "HIS TWO NARRATORS", nodes: narrators });
    return layers;
  }

  /** @param {string} riwayah @returns {IsnadLayer[]} */
  _chainOfNarrator(riwayah) {
    const imam = this.imamOf(riwayah);
    if (!imam) return [];
    const layers = this.topLayers(imam);
    if (!layers.length) return [];
    layers.push({ title: "THE IMAM", nodes: [{ name: imam, arabic: "", detail: "", role: "imam" }] });
    const chain = this._narrators[riwayah];
    if (chain.links.length) {
      layers.push({ title: chain.links.length === 1 ? "THE LINK BETWEEN" : "THE LINKS BETWEEN", nodes: chain.links });
    }
    layers.push({
      title: "THE NARRATOR",
      nodes: [{ name: riwayah.slice(0, riwayah.indexOf(" an ")), arabic: "", detail: "", role: "narrator" }],
    });
    if (chain.students.length) layers.push({ title: "HIS STUDENTS", nodes: chain.students });
    return layers;
  }

  /**
   * One sentence on how a narrator reaches his imam: directly, or through the links between.
   * @param {string} riwayah @returns {string}
   */
  sentence(riwayah) {
    const imam = this.imamOf(riwayah);
    const chain = this._narrators[riwayah];
    if (!imam || !chain) return "";
    const narrator = riwayah.slice(0, riwayah.indexOf(" an "));
    if (!chain.links.length) {
      return `${narrator} read on ${imam} himself, and ${imam}'s chain runs through his teachers to the Companions and to the Prophet ﷺ.`;
    }
    const names = chain.links.map((n) => n.name);
    const path = names.length === 1 ? names[0] : `${names.slice(0, -1).join(", ")} and then ${names[names.length - 1]}`;
    return `${narrator} did not meet ${imam}: the reading reached him through ${path}, and from ${imam} it runs through his teachers to the Companions and to the Prophet ﷺ.`;
  }

  count() {
    return {
      imams: Object.keys(this._imams).length,
      narrators: Object.keys(this._narrators).length,
      companions: this._companions.length,
    };
  }
}

File: packages/quran-engine-py/quran_engine/isnad.py
"""The chains of transmission (isnad) of the Ten Readings. Mirrors src/isnad.js.

From the Prophet \ufdfa down through the Companions who learned from him, the
Successors who taught each imam, the imam himself, the links between him and each of his two
narrators, and the students who carried each narration on.

The links are the standard ones of the classical record: Ibn al-Jazari's al-Nashr and Ghayat
al-Nihayah, al-Dani's al-Taysir, and the turuq of al-Shatibiyyah and al-Durrah. Where a
narrator's students are not listed with confidence, the layer is simply absent.

A chain is returned as LAYERS, one per generation, top (the Prophet) to bottom, which is how a
consumer draws it: a row per layer, connected downward.
"""
from __future__ import annotations
from typing import Optional


class IsnadNode(dict):
    """name, arabic, detail, role."""


class IsnadLayer(dict):
    """title, nodes."""


class Isnad:
    def __init__(self, data: Optional[dict] = None):
        data = data or {}
        self._prophet = IsnadNode(data["prophet"]) if data.get("prophet") else None
        self._companions = [IsnadNode(c) for c in data.get("companions", [])]
        self._imams = data.get("imams", {})
        self._narrators = data.get("narrators", {})

    @property
    def is_loaded(self) -> bool:
        return bool(self._imams)

    def prophet(self) -> Optional[IsnadNode]:
        return self._prophet

    def companions(self) -> list[IsnadNode]:
        """The thirteen Companions the readings are transmitted from."""
        return self._companions

    def imam_keys(self) -> list[str]:
        return list(self._imams)

    def narrator_keys(self) -> list[str]:
        return list(self._narrators)

    def imam(self, imam: str) -> Optional[dict]:
        return self._imams.get(imam)

    def narrator(self, riwayah: str) -> Optional[dict]:
        return self._narrators.get(riwayah)

    def reads_directly(self, riwayah: str) -> bool:
        """Whether a narrator read on his imam himself, with nobody between them."""
        chain = self._narrators.get(riwayah)
        return bool(chain) and not chain["links"]

    def imam_of(self, riwayah: str) -> Optional[str]:
        """The imam a riwayah comes from.

        Read from the corpus, NOT parsed off the tag: four tags name the imam in the Arabic
        genitive ("ad-Duri an Abi Amr") while his key is the nominative ("Abu Amr"), so splitting
        on " an " would resolve those four to nothing.
        """
        chain = self._narrators.get(riwayah)
        imam = chain.get("imam") if chain else None
        return imam if imam in self._imams else None

    def top_layers(self, imam: str) -> list[IsnadLayer]:
        """The Prophet, the Companions his teachers read on, and those teachers."""
        chain = self._imams.get(imam)
        if not chain:
            return []
        layers: list[IsnadLayer] = []
        if self._prophet:
            layers.append(IsnadLayer(title="THE PROPHET", nodes=[self._prophet]))
        if chain["companions"]:
            layers.append(IsnadLayer(title="THE COMPANIONS", nodes=[IsnadNode(c) for c in chain["companions"]]))
        if chain["teachers"]:
            title = "HIS TEACHER" if len(chain["teachers"]) == 1 else "HIS TEACHERS"
            layers.append(IsnadLayer(title=title, nodes=[IsnadNode(t) for t in chain["teachers"]]))
        return layers

    def chain(self, key: str) -> list[IsnadLayer]:
        """A whole chain as layers: a riwayah tag for one narration, an imam key for a reading."""
        k = (key or "").strip()
        if k in self._imams:
            return self._chain_of_imam(k)
        if k in self._narrators:
            return self._chain_of_narrator(k)
        return []

    def _chain_of_imam(self, imam: str) -> list[IsnadLayer]:
        layers = self.top_layers(imam)
        if not layers:
            return []
        layers.append(IsnadLayer(title="THE IMAM",
                                 nodes=[IsnadNode(name=imam, arabic="", detail="", role="imam")]))
        narrators = [
            IsnadNode(name=tag[:tag.find(" an ")], arabic="", detail="", role="narrator")
            for tag in self._narrators if self.imam_of(tag) == imam
        ]
        if narrators:
            layers.append(IsnadLayer(title="HIS TWO NARRATORS", nodes=narrators))
        return layers

    def _chain_of_narrator(self, riwayah: str) -> list[IsnadLayer]:
        imam = self.imam_of(riwayah)
        if not imam:
            return []
        layers = self.top_layers(imam)
        if not layers:
            return []
        layers.append(IsnadLayer(title="THE IMAM",
                                 nodes=[IsnadNode(name=imam, arabic="", detail="", role="imam")]))
        chain = self._narrators[riwayah]
        if chain["links"]:
            title = "THE LINK BETWEEN" if len(chain["links"]) == 1 else "THE LINKS BETWEEN"
            layers.append(IsnadLayer(title=title, nodes=[IsnadNode(n) for n in chain["links"]]))
        layers.append(IsnadLayer(
            title="THE NARRATOR",
            nodes=[IsnadNode(name=riwayah[:riwayah.find(" an ")], arabic="", detail="", role="narrator")]))
        if chain["students"]:
            layers.append(IsnadLayer(title="HIS STUDENTS", nodes=[IsnadNode(n) for n in chain["students"]]))
        return layers

    def sentence(self, riwayah: str) -> str:
        """One sentence on how a narrator reaches his imam: directly, or through the links between."""
        imam = self.imam_of(riwayah)
        chain = self._narrators.get(riwayah)
        if not imam or not chain:
            return ""
        narrator = riwayah[:riwayah.find(" an ")]
        if not chain["links"]:
            return (f"{narrator} read on {imam} himself, and {imam}'s chain runs through his "
                    "teachers to the Companions and to the Prophet ﷺ.")
        names = [n["name"] for n in chain["links"]]
        path = names[0] if len(names) == 1 else ", ".join(names[:-1]) + " and then " + names[-1]
        return (f"{narrator} did not meet {imam}: the reading reached him through {path}, and from "
                f"{imam} it runs through his teachers to the Companions and to the Prophet ﷺ.")

    def count(self) -> dict:
        return {"imams": len(self._imams), "narrators": len(self._narrators),
                "companions": len(self._companions)}
